### PR TITLE
feat(analytics): scroll tracking

### PR DESCRIPTION
**Pull request by annawen1**
### Related Ticket(s)

[Analytics: Add event triggers for scroll tracking #1602](https://github.ibm.com/webstandards/digital-design/issues/1602)

### Description

- Add event trigger for scrolling. Fires every 400px interval (only at the deepest scroll depth - does not fire if user scrolls back up)

### Changelog

**New**

- add scroll event listener
- add `RegisterAnalyticsEvent` in `services` package

**Changed**

- move analytics script to the `preview-head.html` file so that the `ibmStats` object is exposed to the iframe

### CODE REVIEW

File: packages/react/src/components/Masthead/Masthead.js [hunk 1]
```diff
@@ -103,8 +107,27 @@ const Masthead = ({ navigation, ...mastheadProps }) => {
         );
       }
     });
+
```
Comment (jeffchew) [inline packages/react/src/components/Masthead/Masthead.js]: @annawen1 as discussed, we should move this out into an `AnalyticsAPI`, which will have the `registerEvent` method, as well as a `initScrollTracker` method.

File: packages/services/src/services/Analytics/Analytics.js [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+ * ibm.com
+ */
+class AnalyticsAPI {
+  /**
+   * service to fire a stats/metrics event for an action
+   *
+   * @param {object} eventData Object with standard IBM metric event properties and values to send to Coremetrics
+   * @returns {object} JSX object
+   * 
+   * example of eventData object expected:
```
Comment (jeffchew) [inline packages/services/src/services/Analytics/Analytics.js]: @annawen1 can this be placed in an `@example`? How does this generate in JSDocs?

File: packages/react/src/components/Masthead/Masthead.js (hunk 1 above)
Comment (annawen1) [reply packages/react/src/components/Masthead/Masthead.js]: done

File: packages/react/.storybook/manager-head.html [hunk 3]
```diff
@@ -1,20 +1,29 @@
-<meta name="keywords" content="IBM, design, system, Carbon, design system, Bluemix, styleguide, style, guide, components, library, pattern, kit, component, cloud, React, React.js">
-<meta name="description" content="A collection of IBM.com components implemented using React and Carbon Design System.">
+<meta
+  name="keywords"
+  content="IBM, design, system, Carbon, design system, Bluemix, styleguide, style, guide, components, library, pattern, kit, component, cloud, React, React.js"
+/>
+<meta
+  name="description"
+  content="A collection of IBM.com components implemented using React and Carbon Design System."
```
Comment (photodow) [inline packages/react/.storybook/manager-head.html]: Is there anything we need to look at in here, or is this just formatting mostly?

File: packages/services/src/services/Analytics/Analytics.js [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+   * method scroll tracking
+   * Only fires the event at 400px intervals and at the deepest scroll depth
+   *
+   **/
+  static initScrollTracker() {
+    let maxScrollDepth = 0;
+    const scrollAnalytics = root.addEventListener('scroll', () => {
+      let scrollDepth = root.pageYOffset;
+
+      if (scrollDepth % 400 === 0 && scrollDepth > maxScrollDepth) {
```
[the remark this answers is not in the document]
Comment (jeffchew) [reply packages/services/src/services/Analytics/Analytics.js]: I'd mark this as resolved for this PR for now. If there's any changes needed we can address in another PR.

File: packages/services/src/services/Analytics/Analytics.js [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        maxScrollDepth = scrollDepth;
+        this.registerEvent({
+          type: 'element',
+          primaryCategory: 'SCROLL DISTANCE',
+          eventName: scrollDepth,
+          executionPath: root.innerWidth,
```
[the remark this answers is not in the document]
Comment (annawen1) [reply packages/services/src/services/Analytics/Analytics.js]: oh I spoke with Juan EOD yesterday and he said we can get rid of the "grid" and just have the numeric value of the window width

File: packages/react/.storybook/manager-head.html (hunk 3 above)
Comment (annawen1) [reply packages/react/.storybook/manager-head.html]: yep it was just my prettier that reformatted the file after I removed the analytics script from here.

File: packages/react/src/components/DotcomShell/DotcomShell.js [hunk 6]
```diff
@@ -25,6 +26,10 @@ const DotcomShell = ({
   children,
   ...mastheadProps
 }) => {
+  useEffect(() => {
+    AnalyticsAPI.initScrollTracker();
```
[the remark this answers is not in the document]
Comment (annawen1) [reply packages/react/src/components/DotcomShell/DotcomShell.js]: I think for this specific scroll tracking event we discussed only placing it on the `DotcomShell`. If adopters didn't want to use the shell and wanted to pull components a la carte with scroll tracking, they could apply our analytics service themselves. Not sure if every component needs to depend on this analytics service.

File: packages/services/src/services/Analytics/Analytics.js (hunk 2 above)
Comment (annawen1) [reply packages/services/src/services/Analytics/Analytics.js]: done

File: packages/react/src/components/DotcomShell/DotcomShell.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (photodow) [reply packages/react/src/components/DotcomShell/DotcomShell.js]: I guess I don't understand what differentiates the DotcomShell and makes it so special that we only want to track scroll event analytics if that's included on a page vs. from any other component. I thought if you're building an IBM.com page these analytics are required, maybe their not though? If we included a `global.js` for events like these then maybe it's not included in every component, but at least it's one place for users to add all the required things to be on IBM.com

File: packages/services/src/services/Analytics/Analytics.js [hunk 7]
```diff
@@ -0,0 +1,78 @@
+import root from 'window-or-global';
+
+/**
+ * @constant {boolean} scrollTracker determines whether scroll tracking analytics is enabled
+ * @private
+ */
+const scrollTracker = process.env.SCROLL_TRACKING === 'true' || false;
```
Comment (photodow) [inline packages/services/src/services/Analytics/Analytics.js]: I'm not sure I understand why this was added. Why are we allowing them to opt into something like this with an extra step at that? I feel like this is adding complexity to the adopter's UX unnecessarily. I also feel like scroll tracking is something we want on every page, and leaving this door open for every adopter to opt-in doesn't sound like a great strategy for getting that data. Maybe I'm not aligned on the intent here though?

Comment (annawen1) [reply packages/services/src/services/Analytics/Analytics.js]: This was something Juan wanted as well - 

Juan: "that will only be enabled on-demand, meaning it is off by default and will be enabled with an option or config function"

File: packages/react/src/components/DotcomShell/DotcomShell.js (hunk 6 above)
[the remark this answers is not in the document]
Comment (jeffchew) [reply packages/react/src/components/DotcomShell/DotcomShell.js]: I think global.js makes sense, or whatever we decide to call it. I just added a `DDOAPI` that needs to get the version number in the DDO, which is an ask from Juan Leon.

File: packages/services/src/services/Analytics/Analytics.js (hunk 7 above)
Comment (photodow) [reply packages/services/src/services/Analytics/Analytics.js]: Ok, cool. I feel like this hurts the quality of our analytics, but ok. It is what it is I guess. I feel like Juan should be doing this work. 😉haha

**Pull request merged**